Code/GraphMol/FindRings.cpp

```diff
@@ -1098,112 +1096,112 @@ int symmetrizeSSSR(ROMol &mol) {
-      if (std::find(symids.begin(), symids.end(), static_cast<int>(eid)) !=
-          symids.end()) {
+  VECT_INT_VECT bondsssrs, bondextras;
+  RingUtils::convertToBonds(sssrs, bondsssrs, mol);
+  RingUtils::convertToBonds(extras, bondextras, mol);
+
+  //
+  // For each "extra" ring, figure out if it could replace a single
+  // ring in the SSSR. A ring could be swapped out if:
+  //
+  // * They are the same size
+  // * The replacement doesn't remove any bonds from the union of the bonds
+  //   in the SSSR.
+  //
+  // The trick is that we count how many times each bond is covered by the
+  // existing SSSR, and then decrement that count for the removed ring and
+  // increment it for the added ring.
+  //
+  // May miss extra rings that would need to swap two (or three...) rings
+  // to be included.
+
+  // counts of each bond
+  std::unordered_map<int, int> bondCounts;
+  for (const auto& r: bondsssrs) {
+    for (const auto& b: r) {
+      bondCounts[b] += 1;
+    }
+  }
+
+  std::unordered_map<int, int> activeBondCounts;
+  auto extraAtomRing = extras.begin();
+  for (auto& extraBondRing: bondextras) {
+    bool foundAMatch = false;
+    for (auto& ring: bondsssrs) {
+      if (ring.size() != extraBondRing.size()) {
+        continue;
+      } else if (!shareABond(extraBondRing, ring)) {
         continue;
       }
-      exr = bextra[eid];
-#if 0
-          std::cerr<<"      "<<eid<<": ";
-          std::copy(extras[eid].begin(),extras[eid].end(),std::ostream_iterator<int>(std::cerr," "));
-          std::cerr<<"  | ";
-          std::copy(exr.begin(),exr.end(),std::ostream_iterator<int>(std::cerr," "));
-          std::cerr<<std::endl;
-#endif
-      if (ssiz == exr.size()) {
-        // std::cerr<<"        possible"<<std::endl;
-        INT_VECT eunion;
-        Union(nunion, exr, eunion);
-#if 0
-            std::cerr<<"           munion: ";
-            std::copy(munion.begin(),munion.end(),std::ostream_iterator<int>(std::cerr," "));
-            std::cerr<<std::endl;
-            std::cerr<<"           eunion: ";
-            std::copy(eunion.begin(),eunion.end(),std::ostream_iterator<int>(std::cerr," "));
-            std::cerr<<std::endl;
-#endif
-        // now check if the eunion is same as the original union from the SSSRs
-        if (eunion.size() == munion.size()) {
-          // we found a symmetric ring
-          symids.push_back(eid);
-          // std::cerr<<"        keep!"<<std::endl;
+
+      // Remove the SSSR ring bonds and add the
+      // extra ring bonds. If any bonds are missing (count = 0),
+      // the extra ring is not a replacement.
+      activeBondCounts.clear();
+      for (auto& bondID: ring) {
+        activeBondCounts[bondID] = bondCounts[bondID] - 1;
+      }
+
+      for (auto& bondID: extraBondRing) {
+        activeBondCounts[bondID] += 1;
+      }
+
+      bool foundAMatch = true;
+      for (auto& bondIDAndCount: activeBondCounts) {
+        if (bondIDAndCount.second < 1) {
```

Comment:
If I understand this correctly, what you are doing here is the equivalent of checking if any bond that is in ring and not in extraBondRing has count 1 in bondCounts.

In pseudocode it would be something like:

```
for bond in (set(ringBonds) - set(extraRingBonds)):
    if bondCountsBond == 1:
         foundAMatch = false;
         break
```

That might speed up things a little.

Reply:
Did some refactoring after this suggestion. I think it should be theoretically faster, even though it wasn't really measurable.

But it definitely improves the readability and makes the logic plainer.

Reply:
This was a really good insight.

Reply:
Thanks! I am actually surprised that you didn't see it first!

Reply:
Yeah, this whole thing made the code significantly simpler, easier to read, and faster. That's a win all around. Thanks!